Author component reference template and writing guidelines (#1415)
o3de/o3de.org: content/docs/contributing/to-docs/templates/component-reference-doc.md

---
linkTitle: Component Reference Document
title: Writing an O3DE Component Reference Document
description: Guidelines for writing a component reference document for Open 3D Engine (O3DE) using the component reference document template. 
weight: 200
toc: true
---

This topic guides you through the writing process for an **Open 3D Engine (O3DE)** component reference document.

The workflow: 

1. [Determine if your document belongs in the O3DE Component Reference](#does-my-document-belong-in-the-o3de-component-reference). 

> 2. [Download the component reference document template.](#download-the-component-reference-document-template)

Author: I went with sptramer's suggestion but included the fix for consistent verb-pretense in the section titles. 
It now says ## Setting up repository and downloading component template.

All documentation must adhere to the standards outlined in the O3DE [Style Guide](/docs/contributing/to-docs/style-guide/)and [Terminology](/docs/contributing/to-docs/terminology/). If you need any help, reach out to the Documentation and Community Special Interest Group (D&C SIG) on [Discord](https://discord.com/invite/o3de). 

## Does my document belong in the O3DE Component Reference?

First, determine whether or not your document belongs in the [Component Reference](/docs/user-guide/components/reference/) section of the O3DE User Guide. The purpose of the component reference document is exactly that, a page that users can quickly reference when using the component in **O3DE Editor**.

Supplemental information, such as using the component across different workflows, belongs in other documentation. That document can be organized in an appropriate section of the User Guide, such as the [Interactivity and Simulation](/docs/user-guide/interactivity/), [Scripting](/docs/user-guide/scripting/), [Networking](/docs/user-guide/networking/), or [Visualization](/docs/user-guide/visualization/) sections.

> ## Download template

Reviewer: ```suggestion
## Set up repository and download component template
```

Suggest that you add some info linking back to the instructions on setting up for contributing to o3de.org (https://www.o3de.org/docs/contributing/to-docs/o3de-docs-repo-setup/)

In a new tab, open the [component reference template](/files/template-component-reference.md) and save it to your local machine. Open the template in a text editor and refer to this page as you fill it out.

## Writing a component reference document

 The following sections walk you through how to fill in each section of the component reference template. Some sections are optional and may not be needed for your component, so you can remove those sections as needed. 

### Metadata

Like all Markdown (`.md`) files in the O3DE Documentation, your component reference document must contain metadata at the beginning of the file. Refer to [Metdata](/docs/contributing/to-docs/style-guide/metadata/) to help you fill in this section. 

See `template-component-reference.md`:
```yaml
---
linkTitle: <Component Name>
title: <Component Name> Component
description: The <Component Name>, which does <function>, is provided by the <Gem name> in Open 3D Engine (O3DE). # Example

> weight: 200 # Example

Author: I noted "optional" in the comment, and described the default behavior without weight in the guidelines.

**Specifications**:

* Capitalize "Gem".

* Follow the [Links](/docs/contributing/to-docs/style-guide/format/#links) guideline for formatting the link.

See `template-component-reference.md`:
```md
## Provider

[Gem name](_/docs/user-guide/gems/reference/<path-to-gem-docs>_)

```

### Dependencies (Optional)

If your component depends on other components to work, list the names of the dependent components and link it to the corresponding component reference document.

**Specifications**:

* Explain how this component relates to its dependencies, if necessary.

* Follow the [Links](/docs/contributing/to-docs/style-guide/format/#links) guideline for formatting the link.

**Examples**:

* [Diffuse component](/docs/user-guide/components/reference/atom/diffuse-probe-grid/): A straightforward example with one dependency.

* [Light component](/docs/user-guide/components/reference/atom/light/#dependencies): The Light component can depend on different Shape components. This example describes the varying dependency and its criteria.

See `template-component-reference.md`:
```md
## Dependencies

> <!-- (Optional) List the component's depdencies -->

Reviewer 2: ```suggestion
<!-- (Optional) List the component's dependencies -->
```

**Specifications**:

* If you're providing multiple use cases, make sure they are unique from each other. 

* An effective use case provides situational context and clearly states the role of the component in the situation. 

* Write in complete sentences and elaborate on any technical terminology.

**Examples**:

* [PhysX Shape Collider component](/docs/user-guide/components/reference/physx/shape-collider/#use-cases) - Clarifies when to use a PhysX Shape Collider component versus a PhysX Collider component. 

See `template-component-reference.md`:
```md
## Use cases

<!-- (Optional) Write the use cases in which a user may want to use this component. -->

```

### Limitations (Optional)

Write about the limitations this component may have. 

**Specifications**:

* Explain how users can work around a limitation, if possible. 

* Write in complete sentences and elaborate on any technical terminology.

* Don't include any "future work". A better place for that is the feature's roadmap, which can be found in the corresponding [SIG repository](https://www.o3de.org/contribute/#join-a-special-interest-group). 

**Examples**:

> * [Physx Shape Collider component](/docs/user-guide/components/reference/physx/shape-collider/#limitations)

Reviewer 2: ```suggestion
* [PhysX Shape Collider component](/docs/user-guide/components/reference/physx/shape-collider/#limitations)
```

{{< tabs >}} 
{{% tab name="Simple set of properties"%}}

For a component card with a simple set of properties, you only need one property table. 

**Examples**:

* [Box Shape component](http://localhost:1313/docs/user-guide/components/reference/shape/box-shape/): A straightforward example where all properties are displayed at the first level in the component card's content hierarchy.

* [Bloom component](/docs/user-guide/components/reference/atom/bloom/): Even though the Bloom component's properties contains dropdown lists, only one property table is needed. This is because the drop down lists are opened by default and are used to contain an array of values. Another exception is the **Overrides** dropdown list, which isn't used in the user's main workflow. So, only a description of the **Overrides-Enabled Override** property is sufficient. 

See `template-component-reference.md`:
```md

## Properties

![<component name> interface](/images/<path-to-images>)

| Property | Description | Value | default |
| - | - | - | - |
|   |   |   |   |
|   |   |   |   |
|   |   |   |   |

```

{{% /tab %}}
{{% tab name="Complex set of properties"%}}  

> For a component card with a complex set of properties, you may need multiple property tables: base properties, and one for each property group or configuration. The first property table is called "Base properties" and contains all of the properties that are displayed at the first level in the component card's content hierarchy. You can name subsequent property tables after the property group name in the component card or the type of configuration. 

Reviewer 2: You might want to mention when and how **tabs** might be needed in their property documentation, as shown in the Camera Rig component. Then you could link to the section on [tab shortcodes](https://o3de.org/docs/contributing/to-docs/style-guide/format/#tabs) in the style guide for details on how to create them in Markdown.

Author: Added!

* **Description**: Describes what the function does. Be clear and concise, and explain any technical terminology. 
* **Parameter**: What parameters this function accepts, if any. If none, write "None".
* **Return**: What this function returns. 
* **Scriptable**: Whether or not this function is scriptable. Write "Yes" or "No". 

**Specifications**:

* Add a sentence specifying the name of the EBus interface and which components the EBus interface communicates with. (Some groups of components share the same EBus.)

* Separate request functions into the "Request Bus" section and notification functions into the "Notification Bus" section. 

* Place the sections at the end of the page. 

**Examples**:

* [Terrain Physics Heightfield Collider component](/docs/user-guide/components/reference/terrain/terrain-physics-collider/)
* [Vegetation Altitude Filter component](/docs/user-guide/components/reference/vegetation-filters/vegetation-altitude-filter/)
* [Shape Reference component](/docs/user-guide/components/reference/shape/shape-reference/)

See `template-component-reference.md`:
```md

## <EBus interface name>RequestBus

> <!-- (Optional) Example: "Use the following request functions with the <Request EBus name> EBus interface to communicate with <components> in your game." -->

Reviewer 2: To clarify that we mean to refer to *this* component, by name:
```suggestion
<!-- (Optional) Example: "Use the following request functions with the <Request EBus name> EBus interface to communicate with <component name> in your game." -->
```

> | Request Name | Description | Paramter | Return | Scriptable |

Reviewer 2: See comment below about "Request" vs "Method". Also, a spelling error fix.
```suggestion
| Method Name | Description | Parameter | Return | Scriptable |
```

| - | - | - | - | - |
|   |   |   |   |   |
|   |   |   |   |   |
|   |   |   |   |   |

## <EBus interface name>NotificationBus

> <!-- (Optional) Example: "Use the following notification functions with the <Notification EBus name> EBus interface to communicate with <components> in your game." -->

Reviewer 2: ```suggestion
<!-- (Optional) Example: "Use the following notification functions with the <Notification EBus name> EBus interface to communicate with <component name> in your game." -->
```

> | Request Name | Description | Paramter | Return | Scriptable |

Reviewer 2: ```suggestion
| Method Name | Description | Parameter | Return | Scriptable |
```

| - | - | - | - | - |
|   |   |   |   |   |
|   |   |   |   |   |
|   |   |   |   |   |

```

## Where does the component reference document live?

Component reference documentation live in the folder: `o3de.org/docs/user-guide/components/reference/`. They are organized into subfolders that group the components by type. Your component reference document may belong to an existing group, or a new group completely. 

> Feel free to reach out to D&C SIG for assistance. 

Reviewer: ```suggestion
Feel free to reach out to the [O3DE Documentation and Community Special Interest Group](https://github.com/o3de/sig-docs-community) for assistance. 
```

Author: Combined your and willihay's suggestion.
It now says: "Feel free to reach out to sig-docs-community on [Discord](https://discord.com/invite/o3de) or [GitHub](https://github.com/o3de/sig-docs-community) for assistance."

> Feel free to reach out to D&C SIG for assistance. 

Author: Combined your and sptramer's suggestion, and changed Slack to Discord. 
It now says: "Feel free to reach out to sig-docs-community on [Discord](https://discord.com/invite/o3de) or [GitHub](https://github.com/o3de/sig-docs-community) for assistance."

## Storing image files

> The `o3de.org/static/images` folder is structured to reflect the `o3de.org/docs/` folder. So, you must store images for your component reference document in the appropriate component group folder: `o3de.org/static/images/user-guide/components/reference/<group>/`.

Reviewer: Recommend removing the `o3de.org` and using the repo structure here. 

```suggestion
The `static/images` directory is structured to reflect the `content/docs/` directory of the O3DE site repository. So, you must store images for your component reference document in the appropriate component group folder: `static/images/user-guide/components/reference/<group>/`.
```

## Updating the O3DE Component Reference

Remember to update the [Component Reference](/docs/user-guide/components/reference/) section and list your new component reference document. 

> 1. In a text editor, open the file `o3de.org/docs/user-guide/components/reference.md`.

Reviewer: ```suggestion
1. In a text editor, open the file `content/docs/user-guide/components/reference.md` from your clone of the o3de.org repo.
```
